clusterloader2/pkg/measurement/util/controlled_pods_indexer.go

```diff
@@ -40,8 +40,183 @@ type ControlledPodsIndexer struct {
+					cpi.updatePodRefCountLocked(newOwnerUID, 1)
+				}
+			},
+			DeleteFunc: func(obj interface{}) {
+				cpi.UIDLock.Lock()
+				defer cpi.UIDLock.Unlock()
+
+				ownerUID, ownerKind := getControllerInfo(obj)
+				if ownerUID == "" || ownerKind != "ReplicaSet" {
+					return
+				}
+
+				cpi.updatePodRefCountLocked(ownerUID, -1)
+				cpi.clearRSDataIfPossibleLocked(ownerUID)
+			},
+		},
+	)
+	rsInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				cpi.UIDLock.Lock()
+				defer cpi.UIDLock.Unlock()
+
+				cpi.handleIncomingRSAddEventLocked(obj)
+			},
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				cpi.UIDLock.Lock()
+				defer cpi.UIDLock.Unlock()
+
+				oldDeploymentUID, _ := getControllerInfo(oldObj)
+				newDeploymentUID, _ := getControllerInfo(newObj)
+				if oldDeploymentUID == newDeploymentUID {
+					return
+				}
+
+				if oldDeploymentUID != "" {
+					cpi.removeDeploymentDataLocked(oldDeploymentUID, getObjUID(oldObj))
+				}
+				cpi.handleIncomingRSAddEventLocked(newObj)
+			},
+			DeleteFunc: func(obj interface{}) {
+				cpi.UIDLock.Lock()
+				defer cpi.UIDLock.Unlock()
+				rsUID := getObjUID(obj)
+				cpi.rsUIDToState[rsUID].Exists = false
+				cpi.clearRSDataIfPossibleLocked(rsUID)
+			},
+		},
+	)
+
+	return cpi, nil
+}
+
+func getControllerInfo(obj interface{}) (types.UID, string) {
+	metaAccessor, err := meta.Accessor(obj)
+	if err != nil {
+		return "", ""
+	}
+	controller := metav1.GetControllerOf(metaAccessor)
+	if controller == nil {
+		return "", ""
+	}
+	return controller.UID, controller.Kind
+}
+
+func getObjUID(obj interface{}) types.UID {
+	metaAccessor, err := meta.Accessor(obj)
+	if err != nil {
+		return ""
+	}
+	return metaAccessor.GetUID()
+}
+
+func (p *ControlledPodsIndexer) clearRSDataIfPossibleLocked(rsUID types.UID) {
+	state := p.rsUIDToState[rsUID]
+	if state != nil && !state.Exists && state.NumPods == 0 {
+		delete(p.rsUIDToState, rsUID)
+		ownerUID := state.OwnerUID
+		if _, ok := p.deploymentUIDToRsUIDs[ownerUID]; ok {
+			p.removeDeploymentDataLocked(ownerUID, rsUID)
+		}
+	}
+}
+
+func (p *ControlledPodsIndexer) removeDeploymentDataLocked(ownerUID, rsUID types.UID) {
+	delete(p.deploymentUIDToRsUIDs[ownerUID], rsUID)
+	if len(p.deploymentUIDToRsUIDs[ownerUID]) == 0 {
+		delete(p.deploymentUIDToRsUIDs, ownerUID)
+	}
+}
+
+func (p *ControlledPodsIndexer) updatePodRefCountLocked(rsUID types.UID, diff int) {
+	if _, ok := p.rsUIDToState[rsUID]; !ok {
+		p.rsUIDToState[rsUID] = &ReplicaSetState{}
+	}
+	p.rsUIDToState[rsUID].NumPods += diff
+}
+
+func (p *ControlledPodsIndexer) handleIncomingRSAddEventLocked(obj interface{}) {
+	ownerUID, _ := getControllerInfo(obj)
+	rsUID := getObjUID(obj)
+	if _, ok := p.rsUIDToState[rsUID]; !ok {
+		p.rsUIDToState[rsUID] = &ReplicaSetState{}
+	}
+	p.rsUIDToState[rsUID].Exists = true
+	if ownerUID == "" {
+		return
+	}
+	p.rsUIDToState[rsUID].OwnerUID = ownerUID
```

Comment:
Please swap with the if ownerUID to handle case when during update the new owner is nil.

Reply:
Done.